src/ingestion/iam_credential_report.py

```diff
@@ -0,0 +1,151 @@
+    )
+    return iam_session
+
+def get_accounts_list(sf_client):
+    accounts_list = []
+    query = f"select account:Id::number from {AWS_ACCOUNTS_TABLE} where timestamp = (select max(timestamp) from {AWS_ACCOUNTS_TABLE});"
+    res = sf_client.cursor().execute(query).fetchall()
+    for account in res:
+        accounts_list.append(account[0])
+    return accounts_list
+
+
+LOAD_REPORT_LIST_QUERY = f"""
+INSERT INTO {CREDENTIAL_REPORTS_TABLE} (timestamp, report)
+SELECT '{{snapshotclock}}'::timestamp_ltz, PARSE_JSON(column1)
+FROM VALUES {{format_string}}
+"""
+
+LOAD_POLICY_LIST_QUERY = f"""
+INSERT INTO {PASSWORD_POLICY_TABLE} (timestamp, policy)
+SELECT '{{snapshotclock}}'::timestamp_ltz, PARSE_JSON(column1)
+FROM VALUES {{format_string}}
+"""
+
+def get_data_worker(account):
+    iam_session = get_aws_client(account)
+    if iam_session:
+        reports = []
+        reports_json = []
+        global credentials_count
+        global accounts_count
+        iam_client = iam_session.client('iam')
+        iam_client.generate_credential_report()
+        while(1):
+            try:
+                r = iam_client.get_credential_report()
+                break
+            except Exception as e:
+                time.sleep(2)
+        policy = iam_client.get_account_password_policy()['PasswordPolicy']
+        report = r['Content'].decode('utf-8').split('\n')
+        data = [i.split(',') for i in report]
+        reports = [dict(zip(data[0], i)) for i in data[1:]]
+        policy['AccountId'] = account
+        reports_json = [json.dumps({**report, "account_id": account}) for report in reports]
+        sf_client = get_snowflake_client()
+
+        if len(reports):
+            query = LOAD_REPORT_LIST_QUERY.format(
+                snapshotclock=datetime.datetime.utcnow().isoformat(),
+                format_string=", ".join(["(%s)"] * len(reports)))
+            sf_client.cursor().execute(query, reports_json)
+
+        if len(policy):
+            query = LOAD_POLICY_LIST_QUERY.format(
+                snapshotclock=datetime.datetime.utcnow().isoformat(),
+                format_string="(%s)")
+            sf_client.cursor().execute(query, json.dumps(policy))
+
+def get_data(accounts_list):
+    start = datetime.datetime.now()
+    with Pool(processes=4) as pool:
+        pool.map(get_data_worker, accounts_list)
+    end = datetime.datetime.now()
+    print(f"start: {start} end: {end} total: {(end - start).total_seconds()}")
+
+def main():
+    sf_client = get_snowflake_client()
+    current_time = datetime.datetime.now(datetime.timezone.utc)
+    last_time = sf_client.cursor().execute(f'SELECT max(timestamp) FROM {CREDENTIAL_REPORTS_TABLE}').fetchall()[0][0]
+    if last_time is None or (current_time - last_time).total_seconds() > 86400:
+        accounts_list = get_accounts_list(sf_client)
+        get_data(accounts_list)
+    else:
+        print("It's not time yet!")
+
+global_aws_client=get_global_aws_client()
```

Comment:
ditto — move this up into the function it's used and shield main() via `if __module__if __name__ == '__main__'` so it doesn't run on import